### PR TITLE
[13.0][IMP] purchase_security_advanced: add new contacts rule

### PR DESCRIPTION
**Pull request by ChristianSantamaria**
- If the user belongs to group_purchase_own_orders and not to group_purchase_manager, only contacts whose purchase representative is the logged-in user can be read

### CODE REVIEW

File: purchase_security_advanced/__manifest__.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
     """,
     "author": "Solvos",
     "license": "LGPL-3",
-    "version": "13.0.2.0.0",
+    "version": "13.0.2.0.1",
```
Comment (dalonsod) [inline purchase_security_advanced/__manifest__.py]: This is a new feature (restricting access to contacts across the application depending on groups belonging), not a bugfix IMO.

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comment and also please change a little bit commit & PR title moving improving to "add new _contacts_ rule".

In the next comment, I will talck about "sales management extension" for this, that could be complicated...

**Comment by dalonsod** [comment]
And now, we'll talk about sales management extension.

This current implementation fits for a purchase situation, but sometimes becomes too restrictive when `sales_team` is installed (typically when Sales app is available), and couldn't be bypassed because we've added a very restrictive `AND` condition for every contacts (`res.partner`) access.

So the two only options should be the following:

- (1) Relax added AND condition right here depending on `sales_team` groups, making a complex and unified condition, or
- (2) Hooking complete condition, so we could add an inner `OR` if needed with a new addon that inherits from current one.

Due to current technical requirements (we're only calling `user.has_groups` method, not inspecting unknown fields for this addon unless `sales_team` is installed), we should actually inspect if the user belongs or not to a certain group, even if such group doesn't exist (not tested, but very easy to check using shell, calling method with an invented group name and checking whether `False` is returned, or an error is raised), With this assumption, checking sales group belonging doesn't require inheriting from `sales_team`... so we'll select (1) option.

And which should be the new condition?

Let's take a look at Purchase (security OCA addon) and Sales (standard)  groups hierarchy:

*Purchase*:
P0 = "purchase.group_purchase_user"                 # Purchases: all documents
P1 = "purchase_security.group_purchase_own_orders"  # Purchases: own documents
P2 = "purchase.group_purchase_manager"              # Purchase Manager

*Sales*:
S0 = "sales_team.group_sale_salesman"             # Sales: own documents
S1 = "sales_team.group_sale_salesman_all_leads"   # Sales: all documents
S2 = "sales_team.group_sale_manager"              # Sales Manager
             
Hierarchies are different, because purchase adds a minor access group between two full access ones (due to `purchase_security` later existence). So we cannot implement sales and purchase code in the same way.

We should add the new programmatic rule only when a user has restricted access to purchase (and without full access in sales) or to sales (and without full access in purchases), or both. Something like this:

if user has S1 or P2 or (P0 and not P1):
=> full access via sales and/or purchases => don't add any rule 
else user has S0 or P1:
=> limited access by purchase or sales, create  A=[purchase_user_id purchase condition] OR B=[user_id sales condition], when A condition could exist, or B, or both

Please add the improvement in a standalone commit, so we preserve "only purchase groups implementation", at least during this development.

**Comment by dalonsod** [comment]
And now, we'll talk about sales management extension.

This current implementation fits for a purchase situation, but sometimes becomes too restrictive when `sales_team` is installed (typically when Sales app is available), and couldn't be bypassed because we've added a very restrictive `AND` condition for every contacts (`res.partner`) access.

So the two only options should be the following:

* (1) Relax added `AND` condition right here depending on `sales_team` groups, making a complex and unified condition, or
* (2) Hooking complete condition, so we could add an inner `OR` if needed with a new addon that inherits from current one.

Due to current technical requirements (we're only calling `user.has_groups` method, not inspecting unknown fields for this addon unless `sales_team` is installed), we should actually inspect if the user belongs or not to a certain group, even if such group doesn't exist (not tested, but very easy to check using shell, calling method with an invented group name and checking whether False is returned, or an error is raised), With this assumption, checking sales group belonging doesn't require inheriting from `sales_team`... so we'll select (1) option.

And which should be the new condition?

Let's take a look at Purchase (security OCA addon) and Sales (standard) groups hierarchy:

**Purchase**:
P0 = "purchase.group_purchase_user" # Purchases: all documents
P1 = "purchase_security.group_purchase_own_orders" # Purchases: own documents
P2 = "purchase.group_purchase_manager" # Purchase Manager

**Sales**:
S0 = "sales_team.group_sale_salesman" # Sales: own documents
S1 = "sales_team.group_sale_salesman_all_leads" # Sales: all documents
S2 = "sales_team.group_sale_manager" # Sales Manager

Hierarchies are different, because purchase adds a minor access group between two full access ones (due to `purchase_security` later existence). So we cannot implement sales and purchase code in the same way.

We should add the new programmatic rule only when a user has restricted access to purchase (and without full access in sales) or to sales (and without full access in purchases), or both. Something like this:

if user has S1 or P2 or (P0 and not P1):-
=> full access via sales and/or purchases => don't add any rule
else user has S0 or P1:
=> limited access by purchase or sales, create A=[purchase_user_id purchase condition] OR B=[user_id sales condition], when A condition could exist, or B, or both

Please add the improvement in a standalone commit, so we preserve "only purchase groups implementation", at least during this development.

**Comment by dalonsod** [comment]
Looks good, I'll test it ASAP

**Comment by dalonsod** [comment]
Locally tested, I've found the following things to be done:

- **Grant own contact user access**. If a user has restricted access and it's not sales or purchase user if he tries to, e.g. create a PO an error is raised (because the default representative should be himself, and it has no access to himself!), you can check. It could be solved adding two OR clauses within added rule, as sales_team_security does (see https://github.com/OCA/sale-workflow/blob/7b9f061278e8b2e09dbce7ba0a6a8b9b8910f478/sales_team_security/models/ir_rule.py#L34-L39).
- **Add index to `user_id` and `purchase_user_id` for better performance**, as https://github.com/OCA/sale-workflow/blob/7b9f061278e8b2e09dbce7ba0a6a8b9b8910f478/sales_team_security/models/res_partner.py#L14 does.
- **Mark `sales_team_security` as an incompatibility** in manifest (see https://github.com/OCA/account-financial-tools/blob/ba0cc109d54ed9304be180809bf2ef8ab6c2b610/account_asset_management/__manifest__.py#L10). It will help us remembering that issue.

**Comment by ChristianSantamaria** [comment]
Okey, updated in test enviroment, if look good to u then squash commits

**Comment by dalonsod** [comment]
Good, I'll test it ASAP

**Review by dalonsod — APPROVED** [review]
Locally and AluTest tested :+1:, ready to be squashed and merged once a last task is made: there are two minor issues, but we can go on with current implementation. Then, please add a `ROADMAP` with the following content:
- When contacts access is restricted, enable visibility to those contacts whom I'm follower (like OCA's `sales_team_security` does, see `message_partner_ids`).
- Change behavior with non-assigned contacts (they become not accessible, but for `sales_team_security`, in certain conditions, are accesible by everyone).

Last issue links to current "No commercial assigned" workaround currently used by customer (and with this new implementation becomes unneeded).

**Pull request merged**